providers/implementations/kem/ml_kem.c

```diff
@@ -19,214 +19,241 @@
-    if (ctx == NULL)
+    if (v == NULL
+        || (ctx = OPENSSL_zalloc(sizeof(*ctx))) == NULL)
         return NULL;
 
     ctx->libctx = PROV_LIBCTX_OF(provctx);
-
-    debug_print("MLKEMKEM newctx returns %p\n", ctx);
+    ctx->vinfo = v;
     return ctx;
 }
 
-static void mlkem_freectx(void *vctx)
+static void ml_kem_freectx(void *vctx)
 {
-    PROV_MLKEM_CTX *ctx = (PROV_MLKEM_CTX *)vctx;
+    PROV_ML_KEM_CTX *ctx = vctx;
 
-    OPENSSL_free(ctx->entropy);
-    debug_print("MLKEMKEM freectx %p\n", ctx);
+    if (ctx->entropy)
+        OPENSSL_cleanse(ctx->entropy, ML_KEM_RANDOM_BYTES);
     OPENSSL_free(ctx);
 }
 
-static int mlkem_init(void *vctx, int operation, void *vkey, void *vauth,
-                      const OSSL_PARAM params[])
+static int ml_kem_init(void *vctx, int op, void *key,
+                       const OSSL_PARAM params[])
 {
-    PROV_MLKEM_CTX *ctx = (PROV_MLKEM_CTX *)vctx;
-    MLKEM768_KEY *mlkemkey = vkey;
+    PROV_ML_KEM_CTX *ctx = vctx;
 
-    debug_print("MLKEMKEM init %p / %p\n", ctx, mlkemkey);
     if (!ossl_prov_is_running())
         return 0;
+    ctx->key = key;
+    ctx->op = op;
+    return ml_kem_set_ctx_params(vctx, params);
+}
 
-    if (mlkemkey->keytype != MLKEM_KEY_TYPE_768)
-        return 0;
-
-    ctx->key = mlkemkey;
-    ctx->op = operation;
-    ctx->entropy = NULL;
+static int ml_kem_encapsulate_init(void *vctx, void *vkey,
+                                   const OSSL_PARAM params[])
+{
+    ML_KEM_PROVIDER_KEYPAIR *key = vkey;
 
-    if (!mlkem_set_ctx_params(vctx, params))
+    if (key == NULL || !have_keys(key)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_KEY);
         return 0;
-
-    debug_print("MLKEMKEM init OK\n");
-    return 1;
+    }
+    return ml_kem_init(vctx, EVP_PKEY_OP_ENCAPSULATE, key, params);
 }
 
-static int mlkem_encapsulate_init(void *vctx, void *vkey,
-                                  const OSSL_PARAM params[])
+static int ml_kem_decapsulate_init(void *vctx, void *vkey,
+                                   const OSSL_PARAM params[])
 {
-    return mlkem_init(vctx, EVP_PKEY_OP_ENCAPSULATE, vkey, NULL, params);
-}
+    ML_KEM_PROVIDER_KEYPAIR *key = vkey;
 
-static int mlkem_decapsulate_init(void *vctx, void *vkey,
-                                  const OSSL_PARAM params[])
-{
-    return mlkem_init(vctx, EVP_PKEY_OP_DECAPSULATE, vkey, NULL, params);
+    if (key == NULL || key->prvkey == NULL) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_KEY);
+        return 0;
+    }
+    return ml_kem_init(vctx, EVP_PKEY_OP_DECAPSULATE, key, params);
 }
 
-static int mlkem_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+static int ml_kem_set_ctx_params(void *vctx, const OSSL_PARAM params[])
 {
-    PROV_MLKEM_CTX *ctx = (PROV_MLKEM_CTX *)vctx;
+    PROV_ML_KEM_CTX *ctx = vctx;
     const OSSL_PARAM *p;
 
-    debug_print("MLKEMKEM set ctx params %p\n", ctx);
     if (ctx == NULL)
         return 0;
     if (params == NULL)
```

Comment:
```suggestion
    if (ossl_param_is_empty(params))
```

Reply:
Staged, perhaps for a follow-on PR, though could be a mixup here if desirable.